lib/HLSL/DxilLegalizeSampleOffsetPass.cpp

```diff
@@ -0,0 +1,182 @@
+
+    CollectIllegalSamples(illegalOffsets, M);
+
+    if (illegalOffsets.empty())
+      return false;
+
+    TryUnrollLoop(illegalOffsets);
+
+    std::vector<Instruction *> ssaIllegalOffsets;
+    CollectIllegalSamples(ssaIllegalOffsets, M);
+    LegalizeOffsets(ssaIllegalOffsets);
+
+    legacy::PassManager PM;
+    PM.add(createPromoteMemoryToRegisterPass());
+    PM.run(M);
+
+    return true;
+  }
+
+private:
+  void TryUnrollLoop(std::vector<Instruction *> &illegalOffsets);
+  void CollectIllegalSamples(std::vector<Instruction *> &illegalOffsets,
+                             Module &M);
+  void CollectIllegalSamples(std::vector<Instruction *> &illegalOffsets,
+                             Function *F, DXIL::OpCode);
+  void LegalizeOffsets(std::vector<Instruction *> &illegalOffsets);
+};
+
+char DxilLegalizeSampleOffsetPass::ID = 0;
+}
+
+void DxilLegalizeSampleOffsetPass::TryUnrollLoop(
+    std::vector<Instruction *> &illegalOffsets) {
+  std::unordered_set<Function *> funcSet;
+  for (Instruction *I : illegalOffsets) {
+    BasicBlock *BB = I->getParent();
+    funcSet.insert(BB->getParent());
+  }
+
+  for (Function *F : funcSet) {
+    DominatorTreeAnalysis DTA;
+    DominatorTree DT = DTA.run(*F);
+    LoopInfo LI;
+    LI.Analyze(DT);
+
+    bool findOffset = false;
+
+    for (auto loopIt = LI.begin(); loopIt != LI.end(); loopIt++) {
+      Loop *loop = *loopIt;
+      for (Instruction *I : illegalOffsets) {
+        BasicBlock *BB = I->getParent();
+        if (loop->contains(BB)) {
+          findOffset = true;
+          break;
+        }
+      }
+      if (findOffset)
+        break;
+    }
+
+    legacy::FunctionPassManager PM(F->getParent());
+    PM.add(createPromoteMemoryToRegisterPass());
+
+    if (findOffset) {
+      PM.add(createCFGSimplificationPass());
+      PM.add(createLCSSAPass());
+      PM.add(createLoopSimplifyPass());
+      PM.add(createLoopRotatePass());
+      PM.add(createSimpleLoopUnrollPass());
+    }
+    PM.run(*F);
+  }
+}
+
+void DxilLegalizeSampleOffsetPass::CollectIllegalSamples(
+    std::vector<Instruction *> &illegalOffsets, Module &M) {
+  for (Function &F : M.functions()) {
+    if (hlsl::OP::IsDxilOpFunc(&F)) {
+      if (F.user_empty())
+        continue;
+      CallInst *CI = cast<CallInst>(*F.user_begin());
+      DXIL::OpCode opcode = hlsl::OP::GetDxilOpFuncCallInst(CI);
+      switch (opcode) {
+      case DXIL::OpCode::Sample:
+      case DXIL::OpCode::SampleBias:
+      case DXIL::OpCode::SampleCmp:
+      case DXIL::OpCode::SampleCmpLevelZero:
+      case DXIL::OpCode::SampleGrad:
+      case DXIL::OpCode::SampleLevel:
+        CollectIllegalSamples(illegalOffsets, &F, opcode);
+        break;
+      default:
+        continue;
+      }
+    }
+  }
+}
+
+void DxilLegalizeSampleOffsetPass::CollectIllegalSamples(
+    std::vector<Instruction *> &illegalOffsets, Function *F, DXIL::OpCode) {
+  for (User *U : F->users()) {
+    CallInst *CI = cast<CallInst>(U);
+    Value *offset0 =
```

Comment:
Lot of repeated pattern here. How about a function to reduce repetition like

    AddIllegalOffset(std::vector<Instruction *>, CallInst *CI, unsigned operand);